refactor(ratelimit): Return Duration from Ratelimit functions

iequidoo opened the pull request.

iequidoo commented on src/webxdc.rs:
```diff
@@ -111,9 +112,9 @@ pub struct WebxdcInfo {
+    /// Time to wait before calling `sendUpdate()` again since the last call.
     /// Should be exposed to `window.sendUpdateInterval` in JS land.
-    pub send_update_interval: usize,
+    pub send_update_interval: Duration,
```
I hope i don't break things here? This is `Serialize`, but it seems that the true public API is the JSON-RPC one

iequidoo replied:
There's already `test_webxdc()`, but it tests `WebxdcMessageInfo` defined in `deltachat-jsonrpc/src/api/types/webxdc.rs`, not this struct.

Still, this `WebxdcInfo` is public, so maybe there are some other implementations or core bindings using it, that's why i asked.

r10s replied:
also, the reference to "Milliseconds" in the comment is still needed for the part that descriped the usage in js land, which is the whole point of this var :)

so we suddenly need more compexity at "Should be exposed is an integer containing milliseconds to `window.sendUpdateInterval` in JS land."

iequidoo replied:
Ok, i reverted the type change here, but added `_ms` suffix because it's Rust land, not even JSON-RPC API. But in that API let it be without the suffix for consistency with JS land.

iequidoo commented on deltachat-jsonrpc/src/api/types/webxdc.rs:
```diff
@@ -72,7 +75,11 @@ impl WebxdcMessageInfo {
             internet_access,
             self_addr,
-            send_update_interval,
+            send_update_interval: send_update_interval
```
For the future i suggest to add the `_ms` suffix, otherwise one can think it's in seconds. Moreover we have other intervals in API which are indeed in seconds

r10s replied:
i agree in general. 

however here, the value is exposed under the same name in js land, where "ms" are quite usual and not often used as  a suffix.

having different names also adds confusion :)

iequidoo commented:
These jobs:
> Rust tests (macos-latest, 1.82.0)

for some reason are in the `Expected — Waiting for status to be reported` state, in dependabot PRs too.

r10s commented:
to me, it was clear  _before_ this PR what gets to  https://webxdc.org/docs/spec/sendUpdate.html#messaging-layer-limits-for-sendupdate 
and less now.

so, not sure if the refactoring is worth effort and discussion, and possibe bugs. the value is anyway not used inside rust.

if we want that change, i agree, we need to add an explicit test that cffi / jsonrpc really gets the thing as type int with value 20000

from reading the PR this is not clear to me

r10s commented on deltachat-jsonrpc/src/api/types/webxdc.rs:
```diff
@@ -72,7 +75,11 @@ impl WebxdcMessageInfo {
             self_addr,
-            send_update_interval,
+            send_update_interval: send_update_interval
+                .checked_add(Duration::from_micros(999))
+                .context("Overflow occurred")?
+                .as_millis()
```
why is checked_add() needed here? 

to enforce correct rounding from milliseconds from microseconds?  this would deserve a comment.

again, i think, in this case just using a plain integer in rust as well as in js land is way easier, and also easier to understand. 

i agree that Duration is good if it is about calculations, but that is not the case here

iequidoo replied:
I moved this rounding code to `src/webxdc.rs` now. `Duration` is indeed probably not needed in the webxdc code, but it's needed in `Ratelimit` because it's a low-level Rust tool. With `Duration` we can even publish it to crates.io if we want, milliseconds aren't good for a crate interface

iequidoo committed:
refactor(ratelimit): Return Duration from Ratelimit functions

- `Ratelimit` is rather a low-level Rust tool, so it should only use `Duration` and `SystemTime`.
- Rename `Ratelimit::update_interval()` to `min_send_interval()`, the word "update" only occurs in
  `last_update`, but it's a private field.

iequidoo commented:
So, now the changes are limited to the `deltachat-ratelimit`. I only renamed `WebxdcInfo::send_update_interval` to `send_update_interval_ms` in addition because it's not related to JS land directly
